src/mlt++/MltChain.cpp

```diff
@@ -34,6 +36,12 @@ Chain::Chain(Profile &profile, const char *id, const char *service)
     }
 
     mlt_producer source = mlt_factory_producer(profile.get_profile(), id, service);
+    if (source
+        && !strcmp("xml", mlt_properties_get(MLT_PRODUCER_PROPERTIES(source), "mlt_service"))) {
+        mlt_producer_close(source);
+        // Always use a producer-consumer for xml so that the profile can be changed
+        source = mlt_factory_producer(profile.get_profile(), "consumer", service);
```

Comment:
I noticed when recently debugging a related issue in Shotcut that the XML as Clip already gets a consumer producer if the profiles mismatch (a feature of the loader producer). Is that different when using a chain? It is not very desirable to add knowledge about certain modules within the framework proper. It is not a hard rule but rather a goal to minimize it.

Reply:
I noticed this too. When encapsulating XML producer in a chain, the problem becomes that sometimes that producer would be wrapped in consumer producer and sometimes it would not. The chain needs to be able to change the profile on the encapsulated producer. The XML producer does not support this very well because it does not propagate the profile change down to the producers inside of it. Forcing it to always use the producer_consumer works around that problem.

>  It is not very desirable to add knowledge about certain modules within the framework proper.

I agree - and I would prefer a generic way to do this. Another option would be to add a private property to producers called "_mutable_profile". If the profile can be changed (like avformat), then the property would be true. If the profile can not be changed (like xml), the property would be false. That could be the condition to decide if a producer_consumer should be used.